### DISCUSSION TITLE
Injecting React via Content Scripts

### DISCUSSION THREAD

**Discussion by JhaAman** [crxjs/chrome-extension-tools#395] [Q&A]
Hey! I'm wondering if there's a tried and true way to inject content-scripts into the DOM using crxjs. 

I tried this a few days ago and actually got something working:
![CleanShot 2022-05-25 at 14 34 33](https://user-images.githubusercontent.com/7852600/171261865-ed7cad22-6f02-4171-8fae-044efb1d27cb.png)

However, no matter what I do, this no longer works. The way I did it was to use 
```
document.querySelector(".FrSOzf")?.append(root);
```
instead of 
```
document.body.append(root);
```

But it no longer renders at all. Is there a way to reliably get this type of injection done?

**Comment by JhaAman** [comment]
Got it working for a second just now, I'm thinking the issue is related to querySelector returning `null` at first because the CSS class isn't rendered at the beginning. Perhaps HMR reloads the querySelector later when everything _is_ rendered, so it works. 

The question is then, how can I get the querySelector to only run after all the DOM elements have loaded in? Or run similar to a useEffect, where it runs everytime new `div` elements are added/removed? Re: https://stackoverflow.com/questions/43147320/document-queryselector-doesnt-work-in-js-file

**Reply by jacksteamdev** [reply]
I think what you're looking for is a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver). I used it for a project that added buttons to Expedia flight search results. It got complicated fast.

Another less efficient alternative is to check the DOM on an interval. Just make sure you stop the interval when you've got what you want. `querySelector` is surprisingly expensive.

#### Resources

- [javascript - Identify page finished loading content after page was already finished loading - Stack Overflow](https://stackoverflow.com/questions/67646916/identify-page-finished-loading-content-after-page-was-already-finished-loading/67711427#67711427)
- [javascript - Performance of MutationObserver to detect nodes in entire DOM - Stack Overflow](https://stackoverflow.com/questions/31659567/performance-of-mutationobserver-to-detect-nodes-in-entire-dom)
- [javascript - Observe mutations on a target node that doesn't exist yet - Stack Overflow](https://stackoverflow.com/questions/38881301/observe-mutations-on-a-target-node-that-doesnt-exist-yet)

**Reply by williamyeny** [reply]
Hey @JhaAman, I know this is late but I implemented a MutationObserver for the purpose of mounting a React app. You might find it useful:

https://gist.github.com/williamyeny/a18e7184afa3974688af6ce64e50f61b

**Reply by nivethan-me** [reply]
Hi @williamyeny thank you for your gist, i got some idea from your code. but i'm wondering how to do this with crxjs plugin & React. 
basically i want to insert a React component if the element is contentEditable. can you give any sample/info regarding this? 

thank you in advance.
